### PR TITLE
impl(generator): extract schema from Discovery Doc

### PR DESCRIPTION
**Pull request by scotthart**
Part of the work for #10980 

Perform some field checking of the Discovery JSON and add types defined as `schema` into a container of "Types".

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001)

### CODE REVIEW

File: generator/internal/discovery_to_proto_test.cc [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+}
+)""";
+
+  auto parsed_json =
+      nlohmann::json::parse(kDiscoveryDocWithCorrectSchema, nullptr, false);
+  ASSERT_FALSE(parsed_json.is_discarded() || parsed_json.is_null());
+  auto types = ExtractTypesFromSchema(parsed_json);
+  ASSERT_THAT(types, IsOk());
+  EXPECT_THAT(*types, Contains(Key(Eq(std::string("Foo")))));
+  EXPECT_THAT(*types, Contains(Key(Eq(std::string("Bar")))));
```
Comment (dbolduc) [inline generator/internal/discovery_to_proto_test.cc:54]: Consider:

```cc
EXPECT_THAT(*types, UnorderedElementsAre(Key("Foo"), Key("Bar")));
```

File: generator/internal/discovery_to_proto_test.cc [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+}
+)""";
+
+  auto parsed_json =
+      nlohmann::json::parse(kDiscoveryDocMissingSchema, nullptr, false);
+  ASSERT_FALSE(parsed_json.is_discarded() || parsed_json.is_null());
+  auto types = ExtractTypesFromSchema(parsed_json);
+  EXPECT_THAT(types, StatusIs(StatusCode::kInvalidArgument));
+  EXPECT_THAT(types.status().message(),
+              HasSubstr("does not contain schemas element"));
```
Comment (dbolduc) [inline generator/internal/discovery_to_proto_test.cc:69]: Consider:

```cc
  EXPECT_THAT(types, StatusIs(StatusCode::kInvalidArgument,
                              HasSubstr("does not contain schemas element")));
```

File: generator/internal/discovery_to_proto_test.cc [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+}
+)""";
+
+  testing_util::ScopedLog log;
+  auto parsed_json =
+      nlohmann::json::parse(kDiscoveryDocSchemaIdMissing, nullptr, false);
+  ASSERT_FALSE(parsed_json.is_discarded() || parsed_json.is_null());
+  auto types = ExtractTypesFromSchema(parsed_json);
+  EXPECT_THAT(types, StatusIs(StatusCode::kInvalidArgument));
+  EXPECT_THAT(types.status().message(), HasSubstr("schema without id"));
```
Comment (dbolduc) [inline generator/internal/discovery_to_proto_test.cc:89]: ditto

File: generator/internal/discovery_to_proto_test.cc [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+)""";
+
+  testing_util::ScopedLog log;
+  auto parsed_json =
+      nlohmann::json::parse(kDiscoveryDocWithMissingType, nullptr, false);
+  ASSERT_FALSE(parsed_json.is_discarded() || parsed_json.is_null());
+  auto types = ExtractTypesFromSchema(parsed_json);
+  EXPECT_THAT(types, StatusIs(StatusCode::kInvalidArgument));
+  EXPECT_THAT(types.status().message(),
+              HasSubstr("contains non object schema"));
```
Comment (dbolduc) [inline generator/internal/discovery_to_proto_test.cc:112]: ditto

File: generator/internal/discovery_to_proto_test.cc [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+)""";
+
+  testing_util::ScopedLog log;
+  auto parsed_json =
+      nlohmann::json::parse(kDiscoveryDocWithNonObjectSchema, nullptr, false);
+  ASSERT_FALSE(parsed_json.is_discarded() || parsed_json.is_null());
+  auto types = ExtractTypesFromSchema(parsed_json);
+  EXPECT_THAT(types, StatusIs(StatusCode::kInvalidArgument));
+  EXPECT_THAT(types.status().message(),
+              HasSubstr("contains non object schema"));
```
Comment (dbolduc) [inline generator/internal/discovery_to_proto_test.cc:137]: ditto

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001)* status: 0 of 5 files reviewed, 8 unresolved discussions (waiting on @dbolduc)

___
*[`generator/internal/discovery_to_proto.cc` line 79 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPdfmE2OPyZ1tHOmKWJ-r1-79:-NPsyDst7Dj1aot2woi0:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3214400581de364c8d9bb20e1d0d27f3948365ea/generator/internal/discovery_to_proto.cc#L79)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Fun fact: the plural of "schema" is "schemata". I guess Google was not aware.
</blockquote></details>

Done.
___
*[`generator/internal/discovery_to_proto.cc` line 90 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPdfmE2OPyZ1tHOmKWJ-r1-90:-NPsyFTI2sUduu_ZS_7L:bj2y3m3) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3214400581de364c8d9bb20e1d0d27f3948365ea/generator/internal/discovery_to_proto.cc#L90)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

should you just `return internal::InvalidArgumentError(...)` here?

I take it you want to find all errors in the discovery doc in one pass. But do we expect this thing to ever fail?

If we want to find our place in the discovery doc, we could log something like `GCP_LOG(ERROR) << " schema has no id. Last seen id=" << last_id;`
</blockquote></details>

I don't expect it to fail, but if we find errors while sanitizing the input, I'd like to get them all at once. Depending on the issue we'd either adjust the code accordingly, or pass the errors along to the team that owns the discovery document.
___
*[`generator/internal/discovery_to_proto.cc` line 97 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPdfmE2OPyZ1tHOmKWJ-r1-97:-NPsyYEi8RfN0xsimttQ:bs4q9op) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3214400581de364c8d9bb20e1d0d27f3948365ea/generator/internal/discovery_to_proto.cc#L97)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

ditto: Consider just returning here.

(Again, I don't know if we expect these to be failing all the time, and it is our job to send upstream fixes)
</blockquote></details>

Ditto.
___
*[`generator/internal/discovery_to_proto_test.cc` line 54 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPdfmE2OPyZ1tHOmKWL-r1-54:-NPsykwkC3tjhv9Pc61S:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3214400581de364c8d9bb20e1d0d27f3948365ea/generator/internal/discovery_to_proto_test.cc#L54)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Consider:

```cc
EXPECT_THAT(*types, UnorderedElementsAre(Key("Foo"), Key("Bar")));
```
</blockquote></details>

Done.
___
*[`generator/internal/discovery_to_proto_test.cc` line 69 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPdfmE2OPyZ1tHOmKWL-r1-69:-NPsylKm3VAnezVQBPA5:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3214400581de364c8d9bb20e1d0d27f3948365ea/generator/internal/discovery_to_proto_test.cc#L69)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Consider:

```cc
  EXPECT_THAT(types, StatusIs(StatusCode::kInvalidArgument,
                              HasSubstr("does not contain schemas element")));
```
</blockquote></details>

Done.
___
*[`generator/internal/discovery_to_proto_test.cc` line 89 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPdfmE2OPyZ1tHOmKWL-r1-89:-NPsylrV1bexW15yKXqH:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3214400581de364c8d9bb20e1d0d27f3948365ea/generator/internal/discovery_to_proto_test.cc#L89)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[`generator/internal/discovery_to_proto_test.cc` line 112 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPdfmE2OPyZ1tHOmKWL-r1-112:-NPsymAS2nAk8YIgNlLx:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3214400581de364c8d9bb20e1d0d27f3948365ea/generator/internal/discovery_to_proto_test.cc#L112)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[`generator/internal/discovery_to_proto_test.cc` line 137 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPdfmE2OPyZ1tHOmKWL-r1-137:-NPsymiM-UaM0v8T0hVF:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3214400581de364c8d9bb20e1d0d27f3948365ea/generator/internal/discovery_to_proto_test.cc#L137)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

**Review by devbww — COMMENTED** [review]
> Reviewed 3 of 5 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001)* status: all files reviewed, 11 unresolved discussions (waiting on @dbolduc and @scotthart)

___
*[`generator/internal/discovery_to_proto.cc` line 90 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPt0TBa0rRnc9LsM1Rg:-NPt0TBa0rRnc9LsM1Rh:bak9yru) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d570c8b4fe2409c05533d8d1916959b1b4de3ac5/generator/internal/discovery_to_proto.cc#L90)):*
> ```cpp
>   for (auto const& s : schemas) {
>     if (!s.contains("id")) {
>       GCP_LOG(ERROR) << " current schema has no id. last schema with id=" << id;
> ```

Nit: Drop leading space?
___
*[`generator/internal/discovery_to_proto.cc` line 90 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPt0ZIi6j1rk5XL2qTy:-NPt0ZIi6j1rk5XL2qTz:b8ign7z) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d570c8b4fe2409c05533d8d1916959b1b4de3ac5/generator/internal/discovery_to_proto.cc#L90)):*
> ```cpp
>   for (auto const& s : schemas) {
>     if (!s.contains("id")) {
>       GCP_LOG(ERROR) << " current schema has no id. last schema with id=" << id;
> ```

Nit: You might want to consider what to say when `id.empty()`.
___
*[`generator/internal/discovery_to_proto.cc` line 97 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPt0wYu0-SR9j_Nm7IK:-NPt0wYu0-SR9j_Nm7IL:b-5zd5sy) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d570c8b4fe2409c05533d8d1916959b1b4de3ac5/generator/internal/discovery_to_proto.cc#L97)):*
> ```cpp
>     if (!s.contains("type") || s["type"] != "object") {
>       std::string type = s.contains("type") ? s["type"] : "untyped";
>       GCP_LOG(ERROR) << id << " not type:object is " << type;
> ```

Nit: Perhaps "is <type>, not type:object"?
___
*[`generator/internal/discovery_to_proto_test.cc` line 54 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPdfmE2OPyZ1tHOmKWL-r1-54:-NPt1Azk8JXozb9ZZdTh:bo6k2ij) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3214400581de364c8d9bb20e1d0d27f3948365ea/generator/internal/discovery_to_proto_test.cc#L54)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

It looks like you can remove the old expectations now.

**Review by devbww — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001)* status: all files reviewed, 11 unresolved discussions (waiting on @dbolduc and @scotthart)

___
*[`generator/internal/discovery_to_proto.cc` line 97 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPt0wYu0-SR9j_Nm7IK:-NPt1br-CgEUgo6d2eNk:bvufz4c) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d570c8b4fe2409c05533d8d1916959b1b4de3ac5/generator/internal/discovery_to_proto.cc#L97)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Nit: Perhaps "is <type>, not type:object"?
</blockquote></details>

There was a `<type>` in there that was gobbled up.

File: generator/internal/discovery_to_proto.cc [hunk 6]
```diff
@@ -84,21 +84,22 @@ StatusOr<std::map<std::string, DiscoveryTypeVertex>> ExtractTypesFromSchema(
...
+    id = s["id"];
     if (!s.contains("type") || s["type"] != "object") {
       std::string type = s.contains("type") ? s["type"] : "untyped";
       GCP_LOG(ERROR) << id << " not type:object is " << type;
       schemas_all_type_object = false;
       continue;
     }
     types.emplace(
-        std::string(s["id"]),
+        std::string(id),
```
Comment (dbolduc) [inline generator/internal/discovery_to_proto.cc:102]: nit: `s/std::string(id)/id/`

File: generator/internal/discovery_to_proto_test.cc [hunk 7]
```diff
@@ -50,6 +50,7 @@ TEST(ExtractTypesFromSchema, Success) {
   ASSERT_FALSE(parsed_json.is_discarded() || parsed_json.is_null());
   auto types = ExtractTypesFromSchema(parsed_json);
   ASSERT_THAT(types, IsOk());
+  EXPECT_THAT(*types, ::testing::UnorderedElementsAre(Key("Foo"), Key("Bar")));
```
Comment (dbolduc) [inline generator/internal/discovery_to_proto_test.cc:53]: nit: we always try to say `using ::testing::Foo;` outside of the expectations.

File: generator/internal/discovery_to_proto_test.cc [hunk 8]
```diff
@@ -50,6 +50,7 @@ TEST(ExtractTypesFromSchema, Success) {
...
   EXPECT_THAT(*types, Contains(Key(Eq(std::string("Foo")))));
   EXPECT_THAT(*types, Contains(Key(Eq(std::string("Bar")))));
```
Comment (dbolduc) [inline generator/internal/discovery_to_proto_test.cc:55]: obsolete; can delete.

**Review by dbolduc — APPROVED** [review]
LGTM, but @devbww had some comments to address

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001)* status: 3 of 5 files reviewed, 12 unresolved discussions (waiting on @dbolduc and @devbww)

___
*[`generator/internal/discovery_to_proto.cc` line 90 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPt0TBa0rRnc9LsM1Rg:-NPt7iyy1w8mfqhL3jum:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d570c8b4fe2409c05533d8d1916959b1b4de3ac5/generator/internal/discovery_to_proto.cc#L90)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Nit: Drop leading space?
</blockquote></details>

Done.
___
*[`generator/internal/discovery_to_proto.cc` line 90 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPt0ZIi6j1rk5XL2qTy:-NPt7kWt9LDL_9B4ajhi:bqmjwcz) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d570c8b4fe2409c05533d8d1916959b1b4de3ac5/generator/internal/discovery_to_proto.cc#L90)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Nit: You might want to consider what to say when `id.empty()`.
</blockquote></details>

Added.
___
*[`generator/internal/discovery_to_proto.cc` line 97 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPt0wYu0-SR9j_Nm7IK:-NPt7qgB5JuwgCFRndqQ:b-qj92pf) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d570c8b4fe2409c05533d8d1916959b1b4de3ac5/generator/internal/discovery_to_proto.cc#L97)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

There was a `<type>` in there that was gobbled up.
</blockquote></details>

Reworded.
___
*[`generator/internal/discovery_to_proto.cc` line 102 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPdfmE2OPyZ1tHOmKWJ-r2-102:-NPt7skaCTDHLTXK6SLf:b-834n6i) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d570c8b4fe2409c05533d8d1916959b1b4de3ac5/generator/internal/discovery_to_proto.cc#L102)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: `s/std::string(id)/id/`
</blockquote></details>

Oops.
___
*[`generator/internal/discovery_to_proto_test.cc` line 54 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPdfmE2OPyZ1tHOmKWL-r1-54:-NPt7ygL1hvSdf8_mUOF:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/3214400581de364c8d9bb20e1d0d27f3948365ea/generator/internal/discovery_to_proto_test.cc#L54)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It looks like you can remove the old expectations now.
</blockquote></details>

Done.
___
*[`generator/internal/discovery_to_proto_test.cc` line 55 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPdfmE2OPyZ1tHOmKWL-r2-55:-NPt7ywjB6EkYXFIo581:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d570c8b4fe2409c05533d8d1916959b1b4de3ac5/generator/internal/discovery_to_proto_test.cc#L55)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

obsolete; can delete.
</blockquote></details>

Done.

**Review by devbww — COMMENTED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001)* status: all files reviewed, 12 unresolved discussions

___
*[`generator/internal/discovery_to_proto.cc` line 90 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPt0ZIi6j1rk5XL2qTy:-NPu3ao88325GgjaMoFa:b-r3umjw) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d570c8b4fe2409c05533d8d1916959b1b4de3ac5/generator/internal/discovery_to_proto.cc#L90)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Added.
</blockquote></details>

Sorry for not being clearer.  I actually meant when `id.empty()` ... that is, when you enter this `if` on the first loop and don't have a "last schema with id".

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001)* status: all files reviewed, 12 unresolved discussions

___
*[`generator/internal/discovery_to_proto.cc` line 90 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11001#-NPt0ZIi6j1rk5XL2qTy:-NPxUSKj59DgiZWD3Ja_:b-5rs7u3) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d570c8b4fe2409c05533d8d1916959b1b4de3ac5/generator/internal/discovery_to_proto.cc#L90)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Sorry for not being clearer.  I actually meant when `id.empty()` ... that is, when you enter this `if` on the first loop and don't have a "last schema with id".
</blockquote></details>

addressed in PR #11009

**Pull request merged**